### DISCUSSION TITLE
Programmatically change layer visibility with Angular and Google Maps

### DISCUSSION THREAD

**Discussion by florinvirdol** [visgl/deck.gl#5753] [Q&A] [answered]
Hi, guys,
I'm using Deck.gl with Angular 10 and Google Maps. 

I'm struggling to programmatically change a Deck.gl layer visibility (based on clicking a button and NgRx store management - but will not enter into details about this). 

Here is some parts of code that I've tried without luck.
I also had a look at https://github.com/visgl/deck.gl/issues/4571 and [HexagonLayer Example](https://deck.gl/gallery/hexagon-layer) and saw @Pessimistress comments about creating a new layer instance if you want to update a prop.

The `layer2` seems to disappear after the timeout but quickly reappears when i'm interacting (dragging) with the map. And i'm also getting these console errors: 
```
deck: error during initialization of GeoJsonLayer({id: 'earthquakes'}) Error: deck.gl: assertion failed.
deck: error during initialization of ScatterplotLayer({id: 'earthquakes-points'}) Error: deck.gl: assertion failed.
```
![image](https://user-images.githubusercontent.com/2658166/117688886-e2fd7d00-b1c1-11eb-8867-de6d873c3d7e.png)

Any help is really appreciated.
Thank you very much!

```
// some things excluded for clarity

private map!: google.maps.Map;
private deckOverlay: GoogleMapsOverlay;

// these two functions from an Angular service for Map handling

renderLayers(layers: unknown[]) {
  this.deckOverlay = new GoogleMapsOverlay({
	  layers
  });
  this.deckOverlay.setMap(this.map);
}

initMap(mapDiv: ElementRef) {
  this.map = new google.maps.Map(mapDiv.nativeElement, {
	  center: this.center,
	  zoom: this.zoom,
	  styles: this.darkStyle,
  });

  let visibility1 = true;
  let visibility2 = true;
  
  let layer1 = new GeoJsonLayer({
    id: 'earthquakes',
    data: 'https://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/all_month.geojson',
    visible: visibility1,
    filled: true,
    pointRadiusMinPixels: 2,
    pointRadiusMaxPixels: 200,
    opacity: 0.4,
    pointRadiusScale: 0.3,
    getRadius: (f: any) => Math.pow(10, f.properties.mag),
    getFillColor: [255, 70, 30, 180],
    autoHighlight: true,
    transitions: {
      getRadius: {
	      type: 'spring',
	      stiffness: 0.1,
	      damping: 0.15
      },
    }
  });
  
  let layer2 = new GeoJsonLayer({
    id: 'uk-subdivisions',
    data: 'https://vector.maps.elastic.co/files/uk_subdivisions_v1.geo.json',
    visible: visibility2,
    filled: true,
    pointRadiusMinPixels: 2,
    pointRadiusMaxPixels: 200,
    opacity: 0.3,
    pointRadiusScale: 0.3,
    getFillColor: [0, 153, 76],
    getLineColor: [1, 1, 1],
    lineWidthMinPixels: 1,
    autoHighlight: true,
    pickable: true
  });

  this.renderLayers([layer1, layer2]);

  // this is mocked and only for testing purpose
  setTimeout(() => {
    visibility1 = false; // this will be an observable
    this.renderLayers([layer1]);
  }, 10000);
}
```

**Comment by sclebal** [comment]
You have to create a new instance of your layer in every change.

Inside your setTimeout you can do:

```
renderLayers([
     layer1.clone({ visible: false })
])
```

And then deck.gl can detect the change.

Also, you can return your layers (layer1, layer2) from a function calle getLayers. Each time you change a variable, call getLayers (inside setTimeout for example) you will have new instances with the new visible value because visibility1 and 2 are out of the function.

**Reply by florinvirdol** [reply]
Thank you for your response, @Clebal ! 

Unfortunately, i tried your exact suggestion but nothing happens, both layers remain visible. Though, i'm not receiving `deck.gl: assertion failed` console errors. 
I also tried removing `setTimeout` and only let your code, but layer is not hiding.
```
renderLayers([layer1.clone({ visible: false })])
```
Do you have any other thoughts?

By the way, this is the only way in order to toggle visibility? Creating a new instance with `clone` and then in a `renderLayer` function adding a new `GoogleMapsOverlay` and setting it to the map?

```
this.deckOverlay = new GoogleMapsOverlay({
	layersArray
});
this.deckOverlay.setMap(this.map);
```

**Reply by sclebal** [reply]
As Pessimistress said, I think you shouldn't create a new overlay each time you call `renderLayers`. Try this option:

```
renderLayers(layers: unknown[]) {
  if (!this.deckOverlay) {
      this.deckOverlay = new GoogleMapsOverlay({
	      layers
      });
      this.deckOverlay.setMap(this.map);
  } else {
      this.deckOverlay.setProps({ layers });
  }
}
```

**Comment by Pessimistress** [comment]
You cannot change a props object once it is created. To update the `visibility` prop, create a new instance by calling `new GeoJsonLayer({...})`, and render it with `overlay.setProps({layers: [...]})`. 

You are seeing the assertion failure because every time you call `renderLayers` you create a new overlay/Deck canvas, and you are trying to render the same layers in multiple Deck contexts. Layer instances are not meant to be reused. Read https://deck.gl/docs/developer-guide/using-layers#faq

[This example](https://deck.gl/gallery/hexagon-layer) shows how to update layer props from user input.

**Reply by florinvirdol** [reply]
Thank you for your answer!

If you have multiple layers, what is the Deck.gl best practice:

- to have only one `GoogleMapsOverlay` containing all the layers?
- or it's ok to have multiple `GoogleMapsOverlay`s containing different layers, hence adding some complexity by also having Overlay management alongside Layer management?

**Reply by Pessimistress** [reply]
Each `GoogleMapsOverlay` essentially creates a new `Deck` instance, therefore a new canvas, and a new WebGL context. 3D layers can only blend/occlude correctly within the same context. And resources cannot be shared across multiple contexts.

**Comment by florinvirdol** [comment, marked as answer]
Thank you very much, @Pessimistress and @Clebal for your help and suggestions! 

As you suggested, i created only an instance of `GoogleMapsOverlay`, i rendered the layers only once and the used an `updateLayer` function to update an instance of an existing layer by cloning it with the new properties (`this.layers[index] = this.layers[index].clone(newProps);`) and setting overlay properties with the altered layers array (`this.deckOverlay.setProps({ layers: Object.values(this.layers) });`) similar to https://github.com/visgl/deck.gl/issues/4501#issuecomment-615336611.

If helpful for anyone, you can find below the entire fixed code:
```
// some things excluded for clarity

private map!: google.maps.Map;
private layers: Layer[];
private deckOverlay: GoogleMapsOverlay = new GoogleMapsOverlay({ layers: [] });

// these two functions from an Angular service for Map handling

renderLayers(layers: Layer[]) {
  this.deckOverlay.setProps({ layers });
  this.deckOverlay.setMap(this.map);
}

updateLayer(id: string, newProps: unknown) {
  const index= this.layers.findIndex((layer) => layer.id === id);
  if (this.layers[index]) {
	  this.layers[index] = this.layers[index].clone(newProps);
	  this.deckOverlay.setProps({ layers: Object.values(this.layers) });
  }
}

initMap(mapDiv: ElementRef) {
  this.map = new google.maps.Map(mapDiv.nativeElement, {
	  center: this.center,
	  zoom: this.zoom,
	  styles: this.darkStyle,
  });

  let visibility1 = true;
  let visibility2 = true;
  
  let layer1 = new GeoJsonLayer({
    id: 'earthquakes',
    data: 'https://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/all_month.geojson',
    visible: visibility1,
    filled: true,
    pointRadiusMinPixels: 2,
    pointRadiusMaxPixels: 200,
    opacity: 0.4,
    pointRadiusScale: 0.3,
    getRadius: (f: any) => Math.pow(10, f.properties.mag),
    getFillColor: [255, 70, 30, 180],
    autoHighlight: true,
    transitions: {
      getRadius: {
	      type: 'spring',
	      stiffness: 0.1,
	      damping: 0.15
      },
    }
  });
  
  let layer2 = new GeoJsonLayer({
    id: 'uk-subdivisions',
    data: 'https://vector.maps.elastic.co/files/uk_subdivisions_v1.geo.json',
    visible: visibility2,
    filled: true,
    pointRadiusMinPixels: 2,
    pointRadiusMaxPixels: 200,
    opacity: 0.3,
    pointRadiusScale: 0.3,
    getFillColor: [0, 153, 76],
    getLineColor: [1, 1, 1],
    lineWidthMinPixels: 1,
    autoHighlight: true,
    pickable: true
  });

  this.layers = [layer1, layer2];
  this.renderLayers(this.layers);

  // this is mocked and only for testing purpose
  setTimeout(() => {
    visibility1 = false; // this will be an observable
    this.updateLayer(layer1.id, { visible: visibility1 });
  }, 10000);
}
```
